### DISCUSSION TITLE
Drone Height Control Issue: Constrained to Takeoff Height and Unstable Flight Behavior

### DISCUSSION THREAD

**Discussion by matteoviel18** [bitcraze/discussions#740] [Q&A]
Hello everyone,

I am encountering an issue with my drones using the Bitcraze framework and the Swarm module. Currently, I am using the functions takeoff_drone4swarm and takeoff_swarm to initiate the drones' takeoff at a certain height:

```
def takeoff_drone4swarm(scf):
    mc_dict[scf._link_uri].take_off(height=0.5)
    
def takeoff_swarm():
    swarm.parallel_safe(takeoff_drone4swarm)
```

where mc_dict is a dictionary with n_drones MotionCommander objects.

After a while, the following functions are called in my code:

```
def velocity_setpoint_drone4swarm(scf):
    c_dict[scf._link_uri].send_velocity_world_setpoint(
        desired_vx[scf._link_uri],
        desired_vy[scf._link_uri],
        desired_vz[scf._link_uri],
        0) 

def velocity_setpoint_swarm():
    swarm.parallel_safe(velocity_setpoint_drone4swarm)
```

where c_dict is a dictionary with n_drones Commander objects.

The problem is that, trying the code experimentally, I've noticed that my (swarm of) drones seems to be constrained to the height of 0.5 meters as in the takeoff_drone4swarm function specified. 

Specifically, when attempting to execute the velocity desired_vz as defined in velocity_setpoint_drone4swarm, the drone do it but promptly returns back to the 0.5 meters height. Since the velocity_setpoint_swarm function is called more times by means of a ROS callback, this leads to a situation where the drone continuously oscillate up and down in flight.

Any suggestion? Maybe should I use something related to the MotionCommander class in the velocity_setpoint_drone4swarm? For example, it would be better to use the following?

```
def velocity_setpoint_drone4swarm(scf):
    mc_dict[scf._link_uri].start_linear_motion(
        desired_vx[scf._link_uri],
        desired_vy[scf._link_uri],
        desired_vz[scf._link_uri],
        0)
```

Thank you in advance for your help!

Best regards,
Matteo

**Comment by knmcguire** [comment]
Hmmm this is an interesting situation. So if I understand correctly:
1- CF takes off with the motion commander. This is technically giving it a velocity command up which will run for the time necessary to go to this height, then it gives it a zero velocity hover command. If you go through the code that is of a ['TYPE_HOVER' setpoint](https://github.com/bitcraze/crazyflie-lib-python/blob/4c42407a4483e83dd08eb73f0c7017425aada564/cflib/crazyflie/commander.py#L155)
2- CF gets a send velocity world setpoint setpoint (not motion commander) which is a [type velocity setpoint](https://github.com/bitcraze/crazyflie-lib-python/blob/4c42407a4483e83dd08eb73f0c7017425aada564/cflib/crazyflie/commander.py#L121)
it moves but when it stops it goes directly to 0.5

I think that either two things are happening:
1- The different setpoint mixing is done differently in the firmware but I don't think that the firmware doesn't look at type anymore (it's more there for legacy)
2- or Once you initiate the motion commander with a take-off, a separate thread is sending 0.5 height setpoints at a fixed interval, regardless of what else is being sent. 

Looking [at the cflib code
](https://github.com/bitcraze/crazyflie-lib-python/blob/4c42407a4483e83dd08eb73f0c7017425aada564/cflib/positioning/motion_commander.py#L95), I think that indeed there is a thread started by the motion commander that keeps sending a fixed setpoint initialized with takeoff, then it first act upon the few send_velocity_world_setpoint() you sent, and once that is finished, acts on the immediate new setpoint which is the one from the motion commander thread.

So yes, as you suggest, your solution is to use only motion commander functions as that is aware of its thread

**Reply by matteoviel18** [reply]
Thanks :)
